src/diffusers/schedulers/scheduling_amused.py

```diff
@@ -0,0 +1,146 @@
+import math
+from dataclasses import dataclass
+from typing import List, Optional, Tuple, Union
+
+import torch
+
+from ..configuration_utils import ConfigMixin, register_to_config
+from ..utils import BaseOutput
+from .scheduling_utils import SchedulerMixin
+
+
+def gumbel_noise(t, generator=None):
+    device = generator.device if generator is not None else t.device
+    noise = torch.zeros_like(t, device=device).uniform_(0, 1, generator=generator).to(t.device)
+    return -torch.log((-torch.log(noise.clamp(1e-20))).clamp(1e-20))
+
+
+def mask_by_random_topk(mask_len, probs, temperature=1.0, generator=None):
+    confidence = torch.log(probs.clamp(1e-20)) + temperature * gumbel_noise(probs, generator=generator)
+    sorted_confidence = torch.sort(confidence, dim=-1).values
+    cut_off = torch.gather(sorted_confidence, 1, mask_len.long())
+    masking = confidence < cut_off
+    return masking
+
+
+@dataclass
+class AmusedSchedulerOutput(BaseOutput):
+    """
+    Output class for the scheduler's `step` function output.
+
+    Args:
+        prev_sample (`torch.FloatTensor` of shape `(batch_size, num_channels, height, width)` for images):
+            Computed sample `(x_{t-1})` of previous timestep. `prev_sample` should be used as next model input in the
+            denoising loop.
+        pred_original_sample (`torch.FloatTensor` of shape `(batch_size, num_channels, height, width)` for images):
+            The predicted denoised sample `(x_{0})` based on the model output from the current timestep.
+            `pred_original_sample` can be used to preview progress or for guidance.
+    """
+
+    prev_sample: torch.FloatTensor
+    pred_original_sample: torch.FloatTensor = None
+
+
+class AmusedScheduler(SchedulerMixin, ConfigMixin):
+    order = 1
+
+    temperatures: torch.Tensor
+
+    @register_to_config
+    def __init__(
+        self,
+        mask_token_id: int,
+        temperature: Union[int, Tuple[int, int], List[int]] = (2, 0),
+        masking_schedule: str = "cosine",
+    ):
+        ...
+
+    def set_timesteps(self, num_inference_steps: int, device: Union[str, torch.device] = None):
+        self.timesteps = torch.arange(num_inference_steps, device=device).flip(0)
+
+        if isinstance(self.config.temperature, (tuple, list)):
+            self.temperatures = torch.linspace(
+                self.config.temperature[0], self.config.temperature[1], num_inference_steps, device=device
+            )
+        else:
+            self.temperatures = torch.linspace(self.config.temperature, 0.01, num_inference_steps, device=device)
```

Comment:
If `temperature` as just an int, I'd expect the scheduler to have constant temperature not  to decrease to `0.01`

Reply:
yeah, that makes more sense to me + the first path can be configured to get the same result. pinging @patil-suraj  since this is just copy pasted from the orig source https://github.com/huggingface/open-muse/blob/07e54080e9eee01e27f0737c16cb917ad43122e1/muse/modeling_transformer_v2.py#L361-L363

Reply:
In the default maskgit scheduler (as presented in the paper), the `temperature` is always annealed to `0` , cf
https://github.com/google-research/maskgit/blob/main/maskgit/libml/parallel_decode.py#L159

The intuition here is that, at the beginning of the process we use higher temperature, to allow the model to diversly sample tokens, but once an image formed we want to reduce the randomness so we decrease the temperature

Reply:
Also, I think this argument should go into `step` method, it does affect the results quite a bit and would be better if we could easily configure it by passing it to `step` via `pipeline`.

Reply:
move the argument from constructor to set_timesteps arg which is passed in from pipeline